### PR TITLE
Initial port to powerpc bare metal

### PR DESCRIPTION
**Pull request by mikey**
This runs in qemu using:

$ ./qemu-system-ppc64 -M powernv -cpu POWER9 -nographic -bios build/firmware.bin

It can also run in microwatt (new POWER softcore) using GHDL or on an FGPA. Instructions on how to run in GHDL are here:
  https://github.com/antonblanchard/microwatt/blob/master/README.md

Thanks to Anton Blanchard and Jordan Niethe for helping with this port.

### CODE REVIEW

**Comment by dpgeorge** [comment]
Thanks for the contribution.  It's nice that it's a minimal port to start with, makes it easier to review!

I guess MicroPython already runs on PowerPC via the unix port, and here is a real bare-metal port.  I'm just wonder though if this scales to eventually add bare-metal ports to other archs like x86, or if these could all be merged into one "bare-cpu" port.  It depends how much they'd have in common (probably not that much) and whether peripherals will be supported or not.  Are there plans to add peripheral support (eg GPIO, SPI) to this port?

What is the reason to add the `skiboot` code?  Seems like this should be provided by the toolchain?

**Comment by mikey** [comment]
> Thanks for the contribution. It's nice that it's a minimal port to start with, makes it easier to review!

Thanks!

> I guess MicroPython already runs on PowerPC via the unix port, and here is a real bare-metal port. I'm just wonder though if this scales to eventually add bare-metal ports to other archs like x86, or if these could all be merged into one "bare-cpu" port. It depends how much they'd have in common (probably not that much) and whether peripherals will be supported or not. 

The core of the port is mostly architecture specific. head.S, the nlr code, the uart, the linker script would all need to change for a different arch like x86. So I think the core code reuse would be minimal.

> Are there plans to add peripheral support (eg GPIO, SPI) to this port?

I'd like to but at this stage we are concentrating on [microwatt](https://github.com/antonblanchard/microwatt) which doesn't have any I/O as yet (other than the UART). The qemu port was just to make it easier to test (for you and me).

FWIW There is some info on microwatt and micropython here https://lwn.net/Articles/796796/  (see "Demo time") and https://www.youtube.com/watch?v=JdMTLs7EMM0

> What is the reason to add the skiboot code? Seems like this should be provided by the toolchain?

The [skiboot](https://github.com/open-power/skiboot) code was there to get going quicker and with a toolchain that doesn't have the libc libraries compiled in (which makes building cross toolchains easier). I could try to get rid of them if they are a problem.

**Comment by dpgeorge** [comment]
> The [skiboot](https://github.com/open-power/skiboot) code was there to get going quicker and with a toolchain that doesn't have the libc libraries compiled in (which makes building cross toolchains easier). I could try to get rid of them if they are a problem.

I only took a brief look and the main thing that stood out was licensing/copyright of these files being different from core MicroPython.  We try to consolidate such files to `lib/` (eg in a submodule) if they are really necessary.

For general points related to new ports, see also this comment: https://github.com/micropython/micropython/issues/4557#issuecomment-467746514

**Comment by mikey** [comment]
(sorry for the slow update)

> I only took a brief look and the main thing that stood out was licensing/copyright of these files being different from core MicroPython. We try to consolidate such files to lib/ (eg in a submodule) if they are really necessary.

I've updated the code to remove these using more compiler builtins now.

All new files are MIT licensed.

> For general points related to new ports, see also this comment: #4557 (comment)
> 1. are the chips widely available to the public and will they be for the foreseeable future?

You can either use qemu or the microwatt softcore from https://github.com/antonblanchard/microwatt

I've added instructions on how to use qemu in ports/powerpc/README.md

> 2. are development boards easily available?

microwatt is a softcore which can be put on numerous FPGA development boards like the Arty A7, Nexys A7, Nexys Video.

> 3. is the toolchain open source (or at least free to obtain)?

Yes, we are using GCC.

> 4. are there plans to support multiple MCUs/SoCs?

Yes. We support microwatt on various FPGA boards. We'd like to support more IO and soft IP which can be built into SOCs

> 5. will there be someone around for the long term to maintain this port?

Yes me! I've been working on Linux on powerpc for 10+ years:
https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/log/?qt=author&q=neuling

> 6. the license (and copyright) must be appropriate.

All MIT (fixed from initial PR).

Other things I've changed this the initial PR:
1) Added a travis powerpc test
2) Cleaned up the makefile CFLAGS options
3) Removed the divide emulation code (since microwatt now support divide instructions)
4) Rebased on latest master branch (small conflict in nlr.h and updated frozentest.mpy).

File: ports/powerpc/main.c [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+        if (pyexec_event_repl_process_char(c)) {
+            break;
+        }
+    }
+    #else
+    pyexec_friendly_repl();
+    #endif
+    //do_str("print('hello world!', list(x+1 for x in range(10)), end='eol\\n')", MP_PARSE_SINGLE_INPUT);
+    //do_str("for i in range(10):\r\n  print(i)", MP_PARSE_FILE_INPUT);
+    #else
```
Comment (shenki) [inline ports/powerpc/main.c]: Should this commented out code be removed?

File: ports/powerpc/uart_core.c [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+ * THE SOFTWARE.
+ */
+
+#include <unistd.h>
+#include <stdbool.h>
+#include "py/mpconfig.h"
+
+/*
+ * Core UART functions to implement for a port
+ */
```
Comment (shenki) [inline ports/powerpc/uart_core.c]: This comment could go

File: ports/powerpc/uart_core.c [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+
+#if MICROPY_MIN_USE_STM32_MCU
```
Comment (shenki) [inline ports/powerpc/uart_core.c]: This looks unrelated to the ppc port

File: ports/powerpc/uart_core.c [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+typedef struct {
+    volatile uint32_t SR;
+    volatile uint32_t DR;
+} periph_uart_t;
+#define USART1 ((periph_uart_t*)0x40011000)
+#endif
+
+static int potato_console;
```
Comment (shenki) [inline ports/powerpc/uart_core.c]: It would be clearer if the potato and the qemu uarts go in different files.

Also mention what/where the potato uart is (I think it's some IP that microwatt is using?)

**Review by shenki — COMMENTED** [review]
Looks good! I have made some minor comments that are my opinion. If you're unsure then we should defer to the micropython experts.

@dpgeorge, I will also be available to help maintain this port.

File: ports/powerpc/main.c (hunk 1 above)
Comment (mikey) [reply ports/powerpc/main.c]: I'll remove. This was left over from the ports/minimal copy.

File: ports/powerpc/uart_core.c (hunk 2 above)
Comment (mikey) [reply ports/powerpc/uart_core.c]: Yep thanks

File: ports/powerpc/uart_core.c (hunk 3 above)
Comment (mikey) [reply ports/powerpc/uart_core.c]: oops, copy and paste from other ports

File: ports/powerpc/uart_core.c (hunk 4 above)
Comment (mikey) [reply ports/powerpc/uart_core.c]: Good point. 
Yeah potato uart is the uart we use for microwatt. It's from https://github.com/skordal/potato. I'll make that clearer in the code.

**Comment by mikey** [comment]
Updated branch to address @shenki comments.

**Comment by dpgeorge** [comment]
I tested this on Arch Linux which only has a powerpc toolchain in the AUR, so instead I tried using `powerpc64le-linux-musl-cross` (gcc 9.2.1) from https://musl.cc .  It compiled except I needed to change the 3rd arg to `main.c:__assert_fail` from `unsigned int` to `int`.

Is it possible to use `int` here with the Ubuntu toolchain?

It also executed correctly in `qemu-system-ppc64` which is great!  But I did notice it used 100% CPU at an idle prompt.  Is it possible to insert an idle instruction in the UART busy wait loop?

Also, I think it'd be good to put in the README some notes about what toolchain to use and/or where to get it.

File: ports/powerpc/Makefile [hunk 5]
```diff
@@ -0,0 +1,70 @@
+include ../../py/mkenv.mk
+
+CROSS = 0
```
Comment (dpgeorge) [inline ports/powerpc/Makefile]: It looks like this variable is unused...?

File: ports/powerpc/uart_core.c [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+		potato_uart_init();
+	} else {
+		lpc_uart_init();
+	}
+}
+
+// Receive single character
+int mp_hal_stdin_rx_chr(void) {
+    unsigned char c = 0;
+#if MICROPY_MIN_USE_STDOUT
```
Comment (dpgeorge) [inline ports/powerpc/uart_core.c]: I think this can go

File: ports/powerpc/uart_core.c [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+    }
+#endif
+    return c;
+}
+
+// Send string of given length
+void mp_hal_stdout_tx_strn(const char *str, mp_uint_t len) {
+#if MICROPY_MIN_USE_STDOUT
+    int r = write(1, str, len);
+    (void)r;
```
Comment (dpgeorge) [inline ports/powerpc/uart_core.c]: Can probably go

File: ports/powerpc/uart_potato.c [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+
+#define POTATO_CONSOLE_TX		0x00
+#define POTATO_CONSOLE_RX		0x08
+#define POTATO_CONSOLE_STATUS		0x10
+#define   POTATO_CONSOLE_STATUS_RX_EMPTY		0x01
+#define   POTATO_CONSOLE_STATUS_TX_EMPTY		0x02
+#define   POTATO_CONSOLE_STATUS_RX_FULL			0x04
+#define   POTATO_CONSOLE_STATUS_TX_FULL			0x08
+#define POTATO_CONSOLE_CLOCK_DIV	0x18
+#define POTATO_CONSOLE_IRQ_EN		0x20
```
Comment (dpgeorge) [inline ports/powerpc/uart_potato.c]: Please line these up

**Review by dpgeorge — COMMENTED** [review]
For code style, tabs are not used, instead 4 spaces.  And some other things.  To make it easy, just run astyle on the files:
```
$ astyle --options=astyle.fmt *.?
```
where astyle.fmt contains:
```
--style=google
--indent=spaces=4
--attach-closing-while
--add-brackets
--indent-switches
--indent-after-parens
--indent-labels
--indent-preproc-define
--indent-preproc-cond
--pad-header
--unpad-paren
--add-braces
--convert-tabs
```

**Comment by mikey** [comment]
> I tested this on Arch Linux which only has a powerpc toolchain in the AUR, so instead I tried using powerpc64le-linux-musl-cross (gcc 9.2.1) from https://musl.cc . It compiled except I needed to change the 3rd arg to main.c:__assert_fail from unsigned int to int.
> 
> Is it possible to use int here with the Ubuntu toolchain?

This is a difference between the definitions of __assert_fail() between glibc and musl. It's a musl issue AFAICT https://www.openwall.com/lists/musl/2019/03/04/6 

> It also executed correctly in qemu-system-ppc64 which is great! But I did notice it used 100% CPU at an idle prompt. Is it possible to insert an idle instruction in the UART busy wait loop?

Unfortunately without adding support to exceptions, we can't do this bare metal on POWER. It's something we could add in the future. 

> Also, I think it'd be good to put in the README some notes about what toolchain to use and/or where to get it.

Will do and I'll say to avoid musl for now.

File: ports/powerpc/Makefile (hunk 5 above)
Comment (mikey) [reply ports/powerpc/Makefile]: Removing

File: ports/powerpc/uart_core.c (hunk 6 above)
Comment (mikey) [reply ports/powerpc/uart_core.c]: Yep, removing.

File: ports/powerpc/uart_core.c (hunk 7 above)
Comment (mikey) [reply ports/powerpc/uart_core.c]: Yep, removing

File: ports/powerpc/uart_potato.c (hunk 8 above)
Comment (mikey) [reply ports/powerpc/uart_potato.c]: Will do.

**Comment by mikey** [comment]
Pushed update to address commends from @dpgeorge 
1) Fixed formatting
2) Removed some dead code
3) Update README to say avoid musl based libc and give suggestions on which cross compilers to use
4) Included link to microwatt

**Comment by mikey** [comment]
Pushed minor fix for whitespace breakage in ports/powerpc/head.S

**Comment by mikey** [comment]
Repushing as travis seems to have tested 308de49 not d27e5f2.

**Comment by mikey** [comment]
Repushed to fix the travis failure once merged.

File: ports/powerpc/Makefile [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+ifneq ("$(ARCH)", "ppc64")
+ifneq ("$(ARCH)", "ppc64le")
+	CROSS_COMPILE = powerpc64le-linux-
+endif
+endif
+
+INC += -I.
+INC += -I$(TOP)
+INC += -I$(BUILD)
+INC += -Iskiboot
```
Comment (dpgeorge) [inline ports/powerpc/Makefile]: this line can go

File: ports/powerpc/README.md [hunk 10]
```diff
@@ -0,0 +1,40 @@
+# The minimal port
```
Comment (dpgeorge) [inline ports/powerpc/README.md]: should be something like "The PowerPC port"

File: ports/powerpc/README.md [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+
+This port is intended to be a minimal MicroPython port that runs in
```
Comment (dpgeorge) [inline ports/powerpc/README.md]: "... MicroPython port to the PowerPC architecture that runs ..."

File: ports/powerpc/main.c (hunk 1 above)
Comment (dpgeorge) [reply ports/powerpc/main.c]: please remove

File: ports/powerpc/powerpc.lds [hunk 12]
```diff
@@ -0,0 +1,10 @@
+SECTIONS
+{
+        . = 0;
+        _start = .;
+        .head : {
+                KEEP(*(.head))
+        }
+
+        . = 0x1700;
```
Comment (dpgeorge) [inline ports/powerpc/powerpc.lds]: maybe add a comment as to why this number / what this line does?

File: ports/powerpc/uart_core.c (hunk 4 above)
Comment (dpgeorge) [reply ports/powerpc/uart_core.c]: does the selection between lpc and potato need to be dynamic, or can it be configured at compile time? (in the latter case there's no need for these variables)

File: py/nlrpowerpc.c [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+        "ld    3, 0xB8(%0) ;"
+        "mtlr  3 ; "
+        "li    3, 1;"
+        "blr ;"
+        :
+        : "r"(&top->regs)
+        :
+    );
+
+    assert(0); /* should never get here */
```
Comment (dpgeorge) [inline py/nlrpowerpc.c]: can this be replaced with `MP_UNREACHABLE`?

**Review by dpgeorge — COMMENTED** [review]
I tested it with the toolchain from https://toolchains.bootlin.com/ and it works well.

Just a few minor items to address, then it should be good to go in.

File: py/nlrpowerpc.c (hunk 13 above)
Comment (mikey) [reply py/nlrpowerpc.c]: Yep.

(github  won't let me comment on the console question so commenting here)
The console is dynamically selected (not compile time), so we need those variables.

File: ports/powerpc/powerpc.lds (hunk 12 above)
Comment (mikey) [reply ports/powerpc/powerpc.lds]: yep, it for putting it right after the exception vectors. I'll add the comment in the code

File: ports/powerpc/README.md (hunk 11 above)
Comment (mikey) [reply ports/powerpc/README.md]: Oops thanks :-)

File: ports/powerpc/Makefile (hunk 9 above)
Comment (mikey) [reply ports/powerpc/Makefile]: ok, thanks

File: ports/powerpc/README.md (hunk 10 above)
Comment (mikey) [reply ports/powerpc/README.md]: yep

**Comment by mikey** [comment]
New push to address comments from @dpgeorge and rebase to latest master
1) Fixed readme text
2) Removed `do_str` code from main.c
3) documented linker offset
4) use MP_UNREACHABLE in nlr code rather than assert(0)

**Comment by dpgeorge** [comment]
Thank's for updating.  Merged in 079cc940a68b3b3b9d47d4402267393a528a0477 with some fixups to code style, mainly indenting.  Also simplified a bit the Makefile because there's already a rule in py/mkrules.mk to build .S files.

Great contribution, thank you!
